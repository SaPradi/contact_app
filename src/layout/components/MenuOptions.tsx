import { NavLink } from 'react-router-dom';
import  '../styles/menuOptions.css';
import ContactIcon from '../../icons/ContactIcon';
import HomeIcon from '../../icons/HomeIcon';
import HeartIcon from '../../icons/HeartIcon';
import AddIcon from '../../icons/AddIcon';
import CloseIcon from '../../icons/CloseIcon';
import { useDispatch } from 'react-redux';
import { AppDispatch } from '../../store/store';
import { toggleDropdownMenu, toggleFormMenu } from '../../store/menuSlice';



const MenuOptions:React.FC = () => {

  const dispath = useDispatch<AppDispatch>();
  const openForm = ():void =>{
    dispath(toggleFormMenu())
    dispath(toggleDropdownMenu())
  }

  return (
    <div className='menu'>

      <div className='menu__header' >
        <button onClick={()=> dispath(toggleDropdownMenu())} aria-label="close button menu">
          <CloseIcon />
        </button>
      </div>

      <ul className='menu-options'>

        <li className='menu-options__option' id="optionOne" role="option">
          <NavLink to='/' aria-label="Overview page" onClick={()=> dispath(toggleDropdownMenu())}>
            <HomeIcon/>
            <p>Overview</p>
          </NavLink>
        </li>

        <li className="menu-options__option" id="optionTwo" role="option">
          <NavLink  to='/contacts' aria-label="Contacts" onClick={()=> dispath(toggleDropdownMenu())}>
            <ContactIcon/>
            <p>Contacts</p>
          </NavLink>
        </li>

        <li className="menu-options__option" id="optionThree" role="option">
          <NavLink to='/favorites' aria-label="Favorites" onClick={()=> dispath(toggleDropdownMenu())}>
            <HeartIcon/>
            <p>Favorites</p>
          </NavLink>
        </li>

        <li className="menu-options__option" id="optionFour" role="option">
          <button onClick={openForm} className="option__button" id="buttonOptionNew" role="optionBottom" aria-haspopup="true"
              aria-expanded="false" aria-controls="new-menu" aria-label='new contact button'>
              <AddIcon/>
              <p>New</p>
          </button>
        </li>

      </ul>

    </div>
  )
}

export default MenuOptions
