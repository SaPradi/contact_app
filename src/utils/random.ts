export function generateRandomId():number{
  const numDigits = Math.floor(Math.random() * 4) + 1;
  const min = Math.pow(10, numDigits - 1);
  const max = Math.pow(10, numDigits) - 1;
  const id = Math.floor(Math.random() * (max - min + 1)) + min;
  return id;
}


export function generateRandomColor():string{
   const numeroAleatorio = Math.floor(Math.random() * 16777216);
   const colorHex = `#${numeroAleatorio.toString(16).padStart(6, '0')}`;
   return colorHex;
}