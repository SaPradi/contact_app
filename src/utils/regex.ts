export function isValidEmail(email:string):boolean {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
}

export function isNotEmpty(str:string):boolean {
    const notEmptyRegex = /^.*\S.*$/;
    return notEmptyRegex.test(str.trim());
}

export function onlyString(str: string): boolean {
    const stringRegex = /^[a-zA-ZáéíóúÁÉÍÓÚñÑüÜ]+$/;
    return stringRegex.test(str);
}