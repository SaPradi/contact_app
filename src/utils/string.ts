export function extractLetterAvatar(first_name:string,last_name:string):string {

    const firstLetter = first_name.charAt(0).toUpperCase();
    const secondLetter = last_name.charAt(0).toUpperCase();

    const nameAvatarLetter = firstLetter + secondLetter;

    return nameAvatarLetter;

}