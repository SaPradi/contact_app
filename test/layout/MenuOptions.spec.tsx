import React from "react";
import { renderWithProviders } from "../utils/test-utils";
import MenuOptions from '../../src/layout/components/MenuOptions';
import { screen } from '@testing-library/react';
import userEvent from "@testing-library/user-event";


describe('Testing <MenuOptions/>', () => {

    const preloadedState = {
        menu: {
          dropdownMenuVisible: true,
          formMenuVisible:false,
          themeMenuVisible:null
        }
    };


    test('Should match with the snapshot', () => {

        const { container } = renderWithProviders(<MenuOptions />);

        expect(container).toMatchSnapshot();

    })


    test('dispatches toggleDropdownMenu action on close button click', async () => {
        const { store } = renderWithProviders(<MenuOptions />,{preloadedState});

        const closeButton = screen.getByLabelText(/close button menu/i);
        await userEvent.click(closeButton);

        const state = store.getState();
        expect(state.menu.dropdownMenuVisible).toBe(false);
    });


    test('dispatches toggleFormMenu and toggleDropdownMenu actions on new contact button click', async () => {
        const { store } = renderWithProviders(<MenuOptions />,{preloadedState});

        const newContactButton = screen.getByLabelText(/new contact button/i);
        await userEvent.click(newContactButton);

        const state = store.getState();
        // 
        expect(state.menu.dropdownMenuVisible).toBe(false);
        expect(state.menu.formMenuVisible).toBe(true);
    });


})
