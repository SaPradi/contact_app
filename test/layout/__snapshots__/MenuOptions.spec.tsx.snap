// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <MenuOptions/> Should match with the snapshot 1`] = `
<div>
  <div
    class="menu"
  >
    <div
      class="menu__header"
    >
      <button
        aria-label="close button menu"
      >
        <svg
          viewBox="0 0 1216 1312"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1202 1066q0 40-28 68l-136 136q-28 28-68 28t-68-28L608 976l-294 294q-28 28-68 28t-68-28L42 1134q-28-28-28-68t28-68l294-294L42 410q-28-28-28-68t28-68l136-136q28-28 68-28t68 28l294 294l294-294q28-28 68-28t68 28l136 136q28 28 28 68t-28 68L880 704l294 294q28 28 28 68"
          />
        </svg>
      </button>
    </div>
    <ul
      class="menu-options"
    >
      <li
        class="menu-options__option"
        id="optionOne"
        role="option"
      >
        <a
          aria-current="page"
          aria-label="Overview page"
          class="active"
          href="/"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m21.743 12.331l-9-10c-.379-.422-1.107-.422-1.486 0l-9 10a.998.998 0 0 0-.17 1.076c.16.361.518.593.913.593h2v7a1 1 0 0 0 1 1h3a1 1 0 0 0 1-1v-4h4v4a1 1 0 0 0 1 1h3a1 1 0 0 0 1-1v-7h2a.998.998 0 0 0 .743-1.669"
            />
          </svg>
          <p>
            Overview
          </p>
        </a>
      </li>
      <li
        class="menu-options__option"
        id="optionTwo"
        role="option"
      >
        <a
          aria-label="Contacts"
          class=""
          href="/contacts"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.5 6.5C7.5 8.981 9.519 11 12 11s4.5-2.019 4.5-4.5S14.481 2 12 2S7.5 4.019 7.5 6.5M20 21h1v-1c0-3.859-3.141-7-7-7h-4c-3.86 0-7 3.141-7 7v1z"
            />
          </svg>
          <p>
            Contacts
          </p>
        </a>
      </li>
      <li
        class="menu-options__option"
        id="optionThree"
        role="option"
      >
        <a
          aria-label="Favorites"
          class=""
          href="/favorites"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20.205 4.791a5.938 5.938 0 0 0-4.209-1.754A5.906 5.906 0 0 0 12 4.595a5.904 5.904 0 0 0-3.996-1.558a5.942 5.942 0 0 0-4.213 1.758c-2.353 2.363-2.352 6.059.002 8.412L12 21.414l8.207-8.207c2.354-2.353 2.355-6.049-.002-8.416"
            />
          </svg>
          <p>
            Favorites
          </p>
        </a>
      </li>
      <li
        class="menu-options__option"
        id="optionFour"
        role="option"
      >
        <button
          aria-controls="new-menu"
          aria-expanded="false"
          aria-haspopup="true"
          aria-label="new contact button"
          class="option__button"
          id="buttonOptionNew"
          role="optionBottom"
        >
          <svg
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 13H5v-2h6V5h2v6h6v2h-6v6h-2z"
            />
          </svg>
          <p>
            New
          </p>
        </button>
      </li>
    </ul>
  </div>
</div>
`;
